### PR TITLE
Feature: swap swig per nunjucks

### PR DESCRIPTION
**Pull request by renatodeleao**
- Replaces templating engine from deprecated [swig](https://github.com/paularmstrong/swig) to [nunjucks](https://mozilla.github.io/nunjucks/) 
- Fixes security issues pointed at https://github.com/SassDoc/sassdoc-theme-default/issues/96

### CODE REVIEW

**Comment by renatodeleao** [comment]
@pascalduez did you have the chance of taking a look at this?
Cheers.

File: src/nunjucks.js [hunk 1]
```diff
@@ -67,9 +60,22 @@ const maybeYiqContrast = color =>
...
+nunjucksEnv.addFilter('pluralize', pluralize)
+nunjucksEnv.addFilter('unescape', unescape)
+
+// debug
+nunjucksEnv.addGlobal('debug', function () {
+  return this.ctx
+})
+
+/**
+ * Warning: The simple API (above; e.g. nunjucks.render) always uses the configuration from the most recent call to nunjucks.configure. Since this is implicit and can result in unexpected side effects, use of the simple API is discouraged in most cases (especially if configure is used); instead, explicitly create an environment using var env = nunjucks.configure(...) and then call env.render(...) etc.
```
Comment (pascalduez) [inline src/nunjucks.js]: Could you please wrap that comment on several lines?

File: src/nunjucks.js [hunk 2]
```diff
@@ -67,9 +60,22 @@ const maybeYiqContrast = color =>
...
+ *
+ * Investigate why it doens't work like this if I export nunjucksEnv instead
```
Comment (pascalduez) [inline src/nunjucks.js]: Nit: `doesn't`

**Comment by pascalduez** [comment]
Cool, thanks @renatodeleao !

I would need to test this before merging/releasing, as sadly we don't have tests here. 
Do you have a working repo/env at hand?

**Comment by renatodeleao** [comment]
@pascalduez thanks for looking at it.
i've tested it on this repo itself, using theme theme styles (meta testing). 

```jsonc
// try this on your package.json
"scripts": {
    "start": "make && sassdoc ./scss/**/*.scss --theme ./index.js"
}
```

It works, but maybe i'm missing something. I'll try to test it on another repo.

**Comment by renatodeleao** [comment]
@pascalduez did a small test with an [external repo](https://whitesmith.github.io/qnorr-styles). Everything seem to be working.

>EDIT: Doesn't seem to work, I tried on a repo of mine. It's not finding the templates.
Will try do dive in later.

I also tried to install it on another repo via git url

```
yarn add git+ssh://git@github.com:renatodeleao/sassdoc-theme-default.git#feature/swap-swig-per-nunjucks
```

But then I realised that we build files at `.gitignored`. They are pushed to `npm` because of the `files` entry at `package.json`. I added the test commit b8313db, and it works. 

Let me know if it works for you. If yes let me rebase and remove this commit before we merge this.

**Comment by renatodeleao** [comment]
@pascalduez sry for the shameless bump, did you have time to give it another spin?

> Let me know if it works for you. If yes let me rebase and remove this commit before we merge this.

File: src/nunjucks.js [hunk 3]
```diff
@@ -1,14 +1,7 @@
...
-import swigFilters from 'swig/lib/filters'
+import nunjucks from 'nunjucks'
 
-const swig = new Swig()
-export default swig
-
-swigExtras.useFilter(swig, 'split')
-swigExtras.useFilter(swig, 'trim')
-swigExtras.useFilter(swig, 'groupby')
+let nunjucksEnv = nunjucks.configure('views')
```
Comment (pascalduez) [inline src/nunjucks.js]: It should be straighten to `nunjucks.configure(path.resolve(__dirname, '..', 'views'))`.

**Comment by pascalduez** [comment]
@renatodeleao I gave it another try and had the same issue. 

```
» [ERROR] Error: template not found: [...]/node_modules/sassdoc-theme-default/views/documentation/index.html.njk
```
Which is a bit hard to graps since the file is present... But if I recall, nunjucks is a bit picky about this.

Fixed with:
```
nunjucks.configure(path.resolve(__dirname, '..', 'views'))
```

**Comment by renatodeleao** [comment]
@pascalduez fixed the paths! (thx)
But now we have new issue: I've rebased this with master, and now i can't build the theme locally.

### Stacktrace
```
» [ERROR] Template render error: (/Users/leao/GitHub/forks/sassdoc-theme-default/views/documentation/index.html.njk) [Line 2, Column 3]
  Error: Cannot use "in" operator to search for "cross-browser-support" in unexpected types.
```

The problematic SASS, every occurence of sassdoc `@group` fails, being the first one

```scss
/* scss/utils/_mixins.scss */
/// Mixin to handle cross browser keyframes for CSS animations.
/// @group cross-browser-support
/// @param {String} $name - Animation name
@mixin keyframes($name) {
```

This error was introduced by https://github.com/SassDoc/sassdoc-theme-default/pull/103, if i delete the `groupDescription` block from `views/documentation/index.html.njk` it works as expected.

I don't have context of this new feature (besides knowing that introduced group descriptions) but since it's minor patch it should allow previous versions code (like this theme own `scss`, to keep working). Can you shed some light here? 
Thanks!

**Comment by pascalduez** [comment]
@pascalduez I guess it needs `sassdoc-extra` with the `groupDescription` feature to be released and used by the theme. I'll proceed on with this feature, release a minor, then the nunjucks rewrite will be a major. I keep you posted. Thanks for your patience!

**Comment by pascalduez** [comment]
@renatodeleao So the `groupDescription` feature is released.  
`sassdoc@2.6.0`, `sassdoc-theme-default@2.7.0`, `sassdoc-extras@2.5.0`  

Please rebase, update your dependencies and re-test.

**Comment by renatodeleao** [comment]
@pascalduez ahhh makes total sense! thanks for the enlightenment 💡. I'll try to do that asap!

**Comment by renatodeleao** [comment]
@pascalduez all clear and building correctly.

**Comment by pascalduez** [comment]
Thanks a bunch @renatodeleao !

**Comment by pascalduez** [comment]
`sassdoc-theme-default@2.8.1`   
`sassdoc@2.7.0`   

We really need tests in the theme, there was a remaining `raw` filter from Swig throwing an exception.

**Comment by pascalduez** [comment]
Need tests (bis), [Nunjucks is not compatible with node 4](https://travis-ci.org/SassDoc/sassdoc/jobs/523029918), while Sassdoc is. 
This should then basically be a breaking change and a major version. 
Let's see if we get complains :)

**Comment by renatodeleao** [comment]
> We really need tests in the theme, there was a remaining raw filter from Swig throwing an exception.

Damn i miss that one!  (came from after the `groupDescription` feature rebase )

----

> Need tests (bis), [Nunjucks is not compatible with node 4](https://travis-ci.org/SassDoc/sassdoc/jobs/523029918), while Sassdoc is.
> This should then basically be a breaking change and a major version.

Oops... This I must admit I didn't even remember 😅! +1 for tests!
Thanks for patching the PR! :shipit:

**Pull request merged**